pkg/controllers/tagging/tagging_controller.go

```diff
@@ -0,0 +1,190 @@
+// tags to resources.
+type TaggingController struct {
+	controllerOptions options.TaggingControllerOptions
+	kubeClient        clientset.Interface
+	nodeLister        v1lister.NodeLister
+	cloud             *awsv1.Cloud
+
+	// Value controlling TaggingController monitoring period, i.e. how often does TaggingController
+	// check node list. This value should be lower than nodeMonitorGracePeriod
+	// set in controller-manager
+	nodeMonitorPeriod time.Duration
+
+	// A map presenting the node and whether it currently exists
+	currNodes map[string]bool
+
+	// A map representing nodes that were ever part of the cluster
+	totalNodes map[string]*v1.Node
+
+	// Representing the user input for tags
+	tags map[string]string
+
+	// Representing the resources to tag
+	resources []string
+}
+
+// NewTaggingController creates a NewTaggingController object
+func NewTaggingController(
+	nodeInformer coreinformers.NodeInformer,
+	kubeClient clientset.Interface,
+	cloud cloudprovider.Interface,
+	nodeMonitorPeriod time.Duration,
+	tags map[string]string) (*TaggingController, error) {
+
+	awsCloud, ok := cloud.(*awsv1.Cloud)
+	if !ok {
+		err := fmt.Errorf("tagging controller does not support %v provider", cloud.ProviderName())
+		return nil, err
+	}
+
+	tc := &TaggingController{
+		kubeClient:        kubeClient,
+		nodeLister:        nodeInformer.Lister(),
+		cloud:             awsCloud,
+		nodeMonitorPeriod: nodeMonitorPeriod,
+		currNodes:         make(map[string]bool),
+		totalNodes:        make(map[string]*v1.Node),
+		tags:              tags,
+	}
+	return tc, nil
+}
+
+// Run will start the controller to tag resources attached to the cluster
+// and untag resources detached from the cluster.
+func (tc *TaggingController) Run(ctx context.Context) {
+	defer utilruntime.HandleCrash()
+
+	wait.UntilWithContext(ctx, tc.MonitorNodes, tc.nodeMonitorPeriod)
+}
+
+func (tc *TaggingController) MonitorNodes(ctx context.Context) {
+	nodes, err := tc.nodeLister.List(labels.Everything())
+	if err != nil {
+		klog.Fatalf("error listing nodes: %s", err)
+	}
+
+	for k := range tc.currNodes {
+		tc.currNodes[k] = false
+	}
+
+	var nodesToTag []*v1.Node
+	for _, node := range nodes {
+		if _, ok := tc.currNodes[node.GetName()]; !ok {
+			nodesToTag = append(nodesToTag, node)
+		}
+
+		tc.totalNodes[node.GetName()] = node
+		tc.currNodes[node.GetName()] = true
+	}
+	tc.tagNodesResources(nodesToTag)
+
+	var nodesToUntag []*v1.Node
+	for nodeName, existed := range tc.currNodes {
+		if existed == false {
+			nodesToUntag = append(nodesToUntag, tc.totalNodes[nodeName])
+		}
+	}
+	tc.untagNodeResources(nodesToUntag)
+}
+
+// tagNodesResources tag node resources from a list of nodes
+// If we want to tag more resources, modify this function appropriately
+func (tc *TaggingController) tagNodesResources(nodes []*v1.Node) {
+	for _, node := range nodes {
+		nodeTagged := false
+		nodeTagged = tc.tagEc2Instances(node)
```

Comment:
Can we tag the nodes in batches? The EC2 tagging APIs take in a list of resources (I have the same comment in the other file as well) so maybe we should batch here and then make the call? That'll help us avoid the total number of calls we have to make to EC2.

Ref - https://docs.aws.amazon.com/cli/latest/reference/ec2/create-tags.html

Note the following though:
> Constraints: Up to 1000 resource IDs. We recommend breaking up this request into smaller batches.

So we need to arrive at a somewhat optimal number for this one. I am inclined to start with 100 maybe for now (we'll probably know more once we perform some load testing of sorts) but want @nckturner to chime in as well.

Reply:
We could, that was my original thought, but I'd like to have granular control over node error and logging, I assume it would be better to tag node one by one, especially when we already do that in [here](https://github.com/kubernetes/cloud-provider-aws/blob/58bc609b9f79e1283b8896011f1590691604b2aa/pkg/providers/v1/tags.go#L223).

And given that speed of adding nodes to the cluster depends on the speed of bringing up resources (EC2 instance for example), we should have quite a bit leeway in tagging them.

Reply:
I'd like @nckturner to chime in here. It looks like using the Batch functionality will help reduce the number of calls that we make to EC2 as well as help this component scale much better but I don't know if there's a particular reason why we had to do it that way.

One downside to batching these calls is that EC2 will fail the entire batch if even one of the instances is having issues (e.g. it gets terminated before we get to tag it or in general there might be a mismatch between the Kubernetes view of nodes and EC2's view of the underlying instances for some reason).

Reply:
I think we could batch call for each node resources and we would retry for each node only. For now, we only tag EC2 instances but when we expand the controller to more resources, we can simply call create_tags for all tagging resources under a node. That way, if a node fails, we can retry that node only, instead of failing other nodes. Pls let me know what you think.

Reply:
resolved offline